### PR TITLE
Fix build for Mono on Ubuntu

### PR DESCRIPTION
**Pull request by amanda-mitchell**
When I downloaded your project, I was unable to build on my Ubuntu 10.04 VM. Technically, `make` still fails, but I've made significant progress, so I wanted to share what I've got.

Remaining issues:
- `build.proj` refers to a task called `FindLastDirectory`, and this does not appear to be available on my system. This is only used by the testing target, so I was able to continue without it.
- While `ZeroMQ.SimpleTests` works just fine, `ZeroMQ.AcceptanceTests` fails with the following output:

```
Specs in ZeroMQ.AcceptanceTests:

Forwarder, when using forwarder device with full subscription

Unhandled Exception:
ZeroMQ.ZmqSocketException: No such file or directory
  at ZeroMQ.ZmqSocket.Receive (System.Byte[] buffer, SocketFlags flags, System.Int32& size) [0x00000] in <filename unknown>:0 
  at ZeroMQ.ZmqSocket+<Receive>c__AnonStorey10.<>m__28 () [0x00000] in <filename unknown>:0 
  at ZeroMQ.ZmqSocket.ExecuteWithTimeout[Byte[]] (System.Func`1 method, TimeSpan timeout) [0x00000] in <filename unknown>:0 
  at ZeroMQ.ZmqSocket.Receive (System.Byte[] buffer, TimeSpan timeout, System.Int32& size) [0x00000] in <filename unknown>:0 
  at ZeroMQ.SendReceiveExtensions.ReceiveFrame (ZeroMQ.ZmqSocket socket, ZeroMQ.Frame frame, TimeSpan timeout) [0x00000] in <filename unknown>:0 
  at ZeroMQ.SendReceiveExtensions.ReceiveFrame (ZeroMQ.ZmqSocket socket, TimeSpan timeout) [0x00000] in <filename unknown>:0 
  at ZeroMQ.AcceptanceTests.DeviceSpecs.when_using_forwarder_device_with_full_subscription.<context>m__24 (ZeroMQ.ZmqSocket sub) [0x00000] in <filename unknown>:0 
  at ZeroMQ.AcceptanceTests.DeviceSpecs.using_threaded_device`1[ZeroMQ.Devices.ForwarderDevice].<context>m__4C () [0x00000] in <filename unknown>:0 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: ZeroMQ.ZmqSocketException: No such file or directory
  at ZeroMQ.ZmqSocket.Receive (System.Byte[] buffer, SocketFlags flags, System.Int32& size) [0x00000] in <filename unknown>:0 
  at ZeroMQ.ZmqSocket+<Receive>c__AnonStorey10.<>m__28 () [0x00000] in <filename unknown>:0 
  at ZeroMQ.ZmqSocket.ExecuteWithTimeout[Byte[]] (System.Func`1 method, TimeSpan timeout) [0x00000] in <filename unknown>:0 
  at ZeroMQ.ZmqSocket.Receive (System.Byte[] buffer, TimeSpan timeout, System.Int32& size) [0x00000] in <filename unknown>:0 
  at ZeroMQ.SendReceiveExtensions.ReceiveFrame (ZeroMQ.ZmqSocket socket, ZeroMQ.Frame frame, TimeSpan timeout) [0x00000] in <filename unknown>:0 
  at ZeroMQ.SendReceiveExtensions.ReceiveFrame (ZeroMQ.ZmqSocket socket, TimeSpan timeout) [0x00000] in <filename unknown>:0 
  at ZeroMQ.AcceptanceTests.DeviceSpecs.when_using_forwarder_device_with_full_subscription.<context>m__24 (ZeroMQ.ZmqSocket sub) [0x00000] in <filename unknown>:0 
  at ZeroMQ.AcceptanceTests.DeviceSpecs.using_threaded_device`1[ZeroMQ.Devices.ForwarderDevice].<context>m__4C () [0x00000] in <filename unknown>:0 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
```

I looked _very_ briefly at the methods in the stack trace, but there was nothing obvious, and the failure is not yet blocking me, so I left it alone.

### CODE REVIEW

**Comment by jgoz** [comment]
Hi, thank you for taking the time to do this. As you can see, I'm in the bad habit of forgetting to update Mono support.

There are a few issues with this pull request that will need to be fixed before I can accept it:
1. The new `NuGet.exe` you've included here is the full (~600KB) executable as opposed to the bootstrapper (~20KB). The issue with this is that future updates to NuGet would have to be manually updated in this repository, as that is normally handled by the bootstrapper. I don't want to commit the full executable unless it's absolutely necessary, so this is something I will investigate.
2. `FindLastDirectory` is a custom build task included in `src/CustomTasks.targets`. Since `build.proj` imports this, any errors around it may be symptoms of a different problem. Again, this warrants investigation.
3. `.Invoke` is necessary on Windows/.NET; these will need to be wrapped in `#if UNIX` pragmas.

File: src/.nuget/NuGet.targets [hunk 1]
```diff
@@ -23,7 +23,7 @@
 
     <PropertyGroup Condition=" '$(OS)' == 'Windows_NT'">
         <!-- Windows specific commands -->
-        <NuGetToolsPath>$([System.IO.Path]::Combine($(SolutionDir), ".nuget"))</NuGetToolsPath>
+        <NuGetToolsPath>$([System.IO.Path]::Combine($(SolutionDir)\.nuget</NuGetToolsPath>
```
Comment (jgoz) [inline src/.nuget/NuGet.targets]: This will break the build on Windows because it's missing two closing parentheses. It should either be reverted or the unnecessary `Path::Combine` should be removed.

File: src/.nuget/NuGet.targets [hunk 2]
```diff
@@ -69,8 +69,6 @@
     <Target Name="CheckPrerequisites">
         <!-- Raise an error if we're unable to locate nuget.exe  -->
         <Error Condition="'$(DownloadNuGetExe)' != 'true' AND !Exists('$(NuGetExePath)')" Text="Unable to locate '$(NuGetExePath)'" />
-        <SetEnvironmentVariable EnvKey="VisualStudioVersion" EnvValue="$(VisualStudioVersion)" Condition=" '$(VisualStudioVersion)' != '' AND '$(OS)' == 'Windows_NT' " />
-        <SetEnvironmentVariable EnvKey="EnableNuGetPackageRestore" EnvValue="true" />
```
Comment (jgoz) [inline src/.nuget/NuGet.targets]: Why did you remove these?

**Comment by jgoz** [comment]
Also, what version of Mono are you using in your development environment?

File: src/.nuget/NuGet.targets (hunk 2 above)
Comment (amanda-mitchell) [inline src/.nuget/NuGet.targets]: xbuild reported that it was an unavailable task.

**Comment by amanda-mitchell** [comment]
I'm using Mono 3.0, built from source.

File: src/.nuget/NuGet.targets (hunk 1 above)
Comment (amanda-mitchell) [inline src/.nuget/NuGet.targets]: I believe that one was an accidental change. I'll revert it.

File: src/.nuget/NuGet.targets (hunk 2 above)
Comment (jgoz) [inline src/.nuget/NuGet.targets]: These are required on Windows build servers. Can you change it to conditionally exclude these for Mono builds?

**Comment by jgoz** [comment]
Ok, I'll need to test this out on my Mono 2.10.x VM. I suspect most Mono users are still on the stable release.

**Comment by amanda-mitchell** [comment]
I've addressed most of your concerns. I dropped in a NuGet bootstrapper that works on Mono, fixed the Invoke calls, and made Windows-specific tasks conditional.

The `FindLastDirectory` task appears to use a feature that is not yet supported on `xbuild`: I could not find an implementation for `CodeTaskFactory`—or support for MSBuild task factories in general—in the latest Mono codebase.

**Comment by jgoz** [comment]
@david-mitchell Thank you very much for this pull request. I got the original NuGet.exe bootstrapper to work on non-Windows environments and fixed some paths for recent changes I'd made. On my Ubuntu VM, everything appears to work, except for the AcceptanceTests. I'll probably end up writing a custom test harness to avoid the MSpec dependency; from what I can tell, its test runner doesn't play nicely on Mono.

Thanks again!
